lib: Changes made to pfx-lists don't take effect.

NaveenThanikachalam opened the pull request:
Say, more than one sequence of a route-map uses the same prefix-list
in the match clause. After the prefix-list match clause is removed
from any one of the route-map sequences, any further changes made to
the prefix-list doesn't dynamically take effect.

The way the dependency of a route-map on a prefix-list is
maintained currently doesn't take into account the number of
sequences of the same route-map that uses the prefix-list.
Hence, even if one of the sequences of the route-map removes
the prefix-list match clause, the dependency is broken.

A reference counter has been introduced to address this issue.

Signed-off-by: NaveenThanikachalam <nthanikachal@vmware.com>

donaldsharp commented on lib/routemap.c:
```diff
@@ -1663,13 +1676,15 @@ static bool route_map_dep_hash_cmp(const void *p1, const void *p2)
-			(char *)hash_release(dep->dep_rmap_hash, (void *)arg);
-		if (rmap_name) {
-			XFREE(MTYPE_ROUTE_MAP_NAME, rmap_name);
+		dep_data = (struct route_map_dep_data *)hash_release(
```
we do not need to cast a void pointer here

NaveenThanikachalam replied:
Thanks, @donaldsharp. I've addressed this.

donaldsharp commented on lib/routemap.c:
```diff
@@ -1649,7 +1660,9 @@ void route_map_event_hook(void (*func)(route_map_event_t, const char *))
 static bool route_map_rmap_hash_cmp(const void *p1, const void *p2)
 {
-	return (strcmp((const char *)p1, (const char *)p2) == 0);
+	return (strcmp(((const struct route_map_dep_data *)p1)->rname,
```
Please fix this function appropriately.  This is a crash waiting to happen.

NaveenThanikachalam replied:
Thanks, @donaldsharp.
I've addressed this.

donaldsharp commented on lib/routemap.c:
```diff
@@ -1757,7 +1779,13 @@ static int route_map_dep_update(struct hash *dephash, const char *dep_name,
 			dep->this_hash = dephash;
 
-		hash_get(dep->dep_rmap_hash, rname, route_map_name_hash_alloc);
+		dep_data = hash_lookup(dep->dep_rmap_hash, rname);
```
Please create an appropriate `struct route_map_dep_data` and pass it in.  Just passing rname is a crash waiting to happen if someone adds new data to the code.

NaveenThanikachalam replied:
Thanks, @donaldsharp. I've addressed it.

donaldsharp commented on lib/routemap.c:
```diff
@@ -1663,13 +1676,17 @@ static bool route_map_dep_hash_cmp(const void *p1, const void *p2)
+		if (dep_data) {
+			XFREE(MTYPE_ROUTE_MAP_NAME, dep_data->rname);
+			XFREE(MTYPE_ROUTE_MAP_DEP_DATA, dep_data);
+			zlog_debug("CLEARED REFERENCE");
```
Please remove this unguarded debug log message.  No need to spam up log files

donaldsharp commented on lib/routemap.c:
```diff
@@ -1691,6 +1708,12 @@ static void route_map_clear_all_references(char *rmap_name)
+static unsigned int route_map_dep_data_hash_make_key(void *p)
+{
+	struct route_map_dep_data *dep_data = p;
+	return (string_hash_make(dep_data->rname));
```
Line between declerations and function, remove the paranthesis around the return statement

donaldsharp commented on lib/routemap.c:
```diff
@@ -1718,7 +1747,9 @@ static unsigned int route_map_dep_hash_make_key(void *p)
 {
-	char *rmap_name = (char *)bucket->data;
+	struct route_map_dep_data *dep_data =
+				(struct route_map_dep_data *)bucket->data;
```
No need to cast a void poiner

donaldsharp commented on lib/routemap.c:
```diff
@@ -1774,8 +1813,21 @@ static int route_map_dep_update(struct hash *dephash, const char *dep_name,
+		dep_data->refcnt--;
+
+		if (!dep_data->refcnt) {
+			ret_dep_data = (struct route_map_dep_data *)
```
no need to cast a void * pointer returned

donaldsharp commented on lib/routemap.c:
```diff
@@ -1843,9 +1895,13 @@ static struct hash *route_map_get_dep_hash(route_map_event_t event)
+	char *rmap_name = NULL;
 	route_map_event_t type = (route_map_event_t)(ptrdiff_t)data;
 
+	dep_data = (struct route_map_dep_data *)bucket->data;
```
again, no need to cast

donaldsharp commented on lib/routemap.c:
```diff
@@ -1663,13 +1676,17 @@ static bool route_map_dep_hash_cmp(const void *p1, const void *p2)
-		if (rmap_name) {
-			XFREE(MTYPE_ROUTE_MAP_NAME, rmap_name);
+		memset(&tmp_dep_data, 0, sizeof(struct route_map_dep_data));
+		tmp_dep_data.rname = (char *)arg;
```
arg is a void pointer no need to cast

donaldsharp commented on lib/routemap.c:
```diff
@@ -1663,13 +1676,17 @@ static bool route_map_dep_hash_cmp(const void *p1, const void *p2)
+		memset(&tmp_dep_data, 0, sizeof(struct route_map_dep_data));
+		tmp_dep_data.rname = (char *)arg;
+		dep_data = hash_release(dep->dep_rmap_hash,
+					(void *)&tmp_dep_data);
```
no need for a (void *) cast this should be fine.

donaldsharp commented on lib/routemap.c:
```diff
@@ -1699,16 +1722,22 @@ static void *route_map_dep_hash_alloc(void *p)
+
+	dep_data = XCALLOC(MTYPE_ROUTE_MAP_DEP_DATA,
+			   sizeof(struct route_map_dep_data));
+	tmp_dep_data = (struct route_map_dep_data *)p;
```
no need to cast a void * pointer

NaveenThanikachalam commented:
> Let's get a rebase and a force push. I do not really want merges ontop of merges if it can be avoided.
> 
> Finally the commit message is a bit incongruent with the change you made. It applies beyond prefix-lists. It applies to any routemap dependancy data that has a name. You've also fixed community lists, Extended community lists, as lists, Large community lists, calls and filters. Let's call that out in the commit message

Thanks Donald. There seems to be a problem in forcing a commit since the head has moved too far forward. In order to avoid any risks, I'll pull this PR down.
I've addressed your comments and created a new PR. Please find the link below:
https://github.com/FRRouting/frr/pull/4345
